x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/PrecompiledCharMapNormalizer.java

```diff
@@ -155,40 +155,43 @@ String normalize(String str) {
         // break points from there. But, this seemed the easiest way for now
         //
         // Keep in mind, these break points aren't necessarily surrogate pairs, but also codepoints that contain a combining mark
+        byte[] strBytes = str.getBytes(StandardCharsets.UTF_8);
+        char[] strChars = str.toCharArray();
+        int[] strCp = str.codePoints().toArray();
         BreakIterator b = BreakIterator.getCharacterInstance(Locale.ROOT);
         b.setText(str);
         int start = b.first();
-        // If we knew the utf-8 length ahead of time (and iterated over the bytes in the appropriate chunks)
-        // we could pre-populate the known length here.
         BytesRefBuilder strBuilder = new BytesRefBuilder();
+        strBuilder.grow(strBytes.length);
+        int bytePos = 0;
         for (int end = b.next(); end != BreakIterator.DONE; start = end, end = b.next()) {
-            // TODO: It would be awesome if we could translate these starts and ends to byte positions, if we could performance would be
-            // dramatically improved
-            String unicodeStr = str.substring(start, end);
-            byte[] unicode = unicodeStr.getBytes(StandardCharsets.UTF_8);
+            int byteLen = 0;
+            int numCp = str.codePointCount(start, end);
+            for (int i = start; i < numCp + start; i++) {
+                byteLen += numUtf8Bytes(strCp[i]);
```

Comment:
I think there's a problem here.  It's assuming that `start` equals the number of code points that have been processed previously.  But actually `start` stores the number of UTF-16 characters that have been processed previously.  They might be the same, but they might not.

I think in this line `start` needs to be `str.codePointCount(0, start)`.  But it's probably more efficient to add an extra variable that keeps a running count of that, similar to what `bytePos` does for the bytes.